### PR TITLE
Selection Range support. Part of Microsoft/vscode#65925

### PR DESCRIPTION
**Pull request by octref**

### CODE REVIEW

File: src/test/selectionRange.test.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	const document = TextDocument.create('test://foo/bar.html', 'html', 1, content);
+	const actualRanges = getApplicableRanges(document, document.positionAt(offset));
+
+	message += `\n${JSON.stringify(actualRanges)} should equal to ${JSON.stringify(expected)}`;
+	assert.deepEqual(actualRanges, expected, message);
+}
+
+suite('HTML SelectionRange', () => {
+	test('Basic', () => {
+		assertRanges('<div|>foo</div>', [[1, 4], [0, 14]]);
```
Comment (aeschli) [inline src/test/selectionRange.test.ts]: shouldn't [1,4] be [1,3]?
For readability, maybe format the expected ranges like `[ "1:div", "0:<div>foo</div>"]`

File: src/test/selectionRange.test.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	} else {
+		content = lines;
+	}
+	let message = `Test ${content}`;
+
+	let offset = content.indexOf('|');
+	content = content.substr(0, offset) + content.substr(offset + 1);
+
+	const document = TextDocument.create('test://foo/bar.html', 'html', 1, content);
+	const actualRanges = getApplicableRanges(document, document.positionAt(offset));
```
Comment (aeschli) [inline src/test/selectionRange.test.ts]: Ideally write the tests against the public API (htmlLanguageService.getSelectionRanges), that way we are sure it works end-to-end and the implementation can be changed freely without impacting any tests.

File: src/test/selectionRange.test.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		assertRanges('<div>|foo</div>', [[5, 8], [0, 14]]);
+		assertRanges('<div>f|oo</div>', [[5, 8], [0, 14]]);
+		assertRanges('<div>foo|</div>', [[5, 8], [0, 14]]);
+
+		assertRanges('<div>foo<|/div>', [[0, 14]]);
+
+		assertRanges('<div>foo</|div>', [[10, 13], [0, 14]]);
+		assertRanges('<div>foo</di|v>', [[10, 13], [0, 14]]);
+		assertRanges('<div>foo</div|>', [[10, 13], [0, 14]]);
+	});
```
Comment (aeschli) [inline src/test/selectionRange.test.ts]: also test self closed tags (`<div foo="sdsdd"/>`)

File: src/test/selectionRange.test.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+	test('Unquoted Attribute Value', () => {
+		assertRanges('<div class=|foo>foo</div>', [[11, 14], [5, 14], [1, 14], [0, 24]]);
+	});
+
+	test('Multiple Attribute Value', () => {
+		assertRanges('<div class="foo" id="|bar">foo</div>', [[21, 24], [20, 25], [17, 25], [1, 25], [0, 35]]);
+	});
+});
```
Comment (aeschli) [inline src/test/selectionRange.test.ts]: test comments and doctypes as well

**Comment by aeschli** [comment]
Nice work!

**Comment by octref** [comment]
@aeschli Thanks! Addressed all your comments.

**Pull request merged**